# Exercise 02 -  OLAP Cubes - Slicing and Dicing

All the databases table in this demo are based on public database samples and transformations
- `Sakila` is a sample database created by `MySql` [Link](https://video.udacity-data.com/topher/2021/August/61120e06_pagila-3nf/pagila-3nf.png)
- The postgresql version of it is called `Pagila` [Link](https://github.com/devrimgunduz/pagila)
- The facts and dimension tables design is based on O'Reilly's public dimensional modelling tutorial schema [Link](https://video.udacity-data.com/topher/2021/August/61120d38_pagila-star/pagila-star.png)

Start by creating and connecting to the database by running the cells below.

In [1]:
!PGPASSWORD=student createdb -h 127.0.0.1 -U student pagila_star
!PGPASSWORD=student psql -q -h 127.0.0.1 -U student -d pagila_star -f Data/pagila-star.sql

 set_config 
------------
 
(1 row)

 setval 
--------
    200
(1 row)

 setval 
--------
    605
(1 row)

 setval 
--------
     16
(1 row)

 setval 
--------
    600
(1 row)

 setval 
--------
    109
(1 row)

 setval 
--------
    599
(1 row)

 setval 
--------
      1
(1 row)

 setval 
--------
      1
(1 row)

 setval 
--------
      1
(1 row)

 setval 
--------
      1
(1 row)

 setval 
--------
  16049
(1 row)

 setval 
--------
   1000
(1 row)

 setval 
--------
   4581
(1 row)

 setval 
--------
      6
(1 row)

 setval 
--------
  32098
(1 row)

 setval 
--------
  16049
(1 row)

 setval 
--------
      2
(1 row)

 setval 
--------
      2
(1 row)



### Connect to the local database where Pagila is loaded

In [2]:
import sql
%load_ext sql

DB_ENDPOINT = "127.0.0.1"
DB = 'pagila_star'
DB_USER = 'student'
DB_PASSWORD = 'student'
DB_PORT = '5432'

# postgresql://username:password@host:port/database
conn_string = "postgresql://{}:{}@{}:{}/{}" \
                        .format(DB_USER, DB_PASSWORD, DB_ENDPOINT, DB_PORT, DB)

print(conn_string)
%sql $conn_string

postgresql://student:student@127.0.0.1:5432/pagila_star


'Connected: student@pagila_star'

### Star Schema

<img src="pagila-star.png" width="50%"/>

# Start with a simple cube
TODO: Write a query that calculates the revenue (sales_amount) by day, rating, and city. Remember to join with the appropriate dimension tables to replace the keys with the dimension labels. Sort by revenue in descending order and limit to the first 20 rows. The first few rows of your output should match the table below.

In [3]:
%%time
%%sql

SELECT
    dt.day,
    dm.rating,
    dc.city,
    sum(sales_amount) as revenue
FROM factsales fs
JOIN dimDate dt ON (fs.date_key = dt.date_key)
JOIN dimMovie dm ON (fs.movie_key = dm.movie_key)
JOIN dimCustomer dc on (fs.customer_key = dc.customer_key)
GROUP BY dt.day, dm.rating, dc.city
ORDER BY 4 DESC


 * postgresql://student:***@127.0.0.1:5432/pagila_star
14167 rows affected.
CPU times: user 26.6 ms, sys: 2.85 ms, total: 29.5 ms
Wall time: 57.1 ms


[(30, 'G', 'San Bernardino', Decimal('24.97')),
 (30, 'NC-17', 'Apeldoorn', Decimal('23.95')),
 (21, 'NC-17', 'Belm', Decimal('22.97')),
 (21, 'G', 'Citt del Vaticano', Decimal('21.97')),
 (28, 'R', 'Mwanza', Decimal('21.97')),
 (30, 'PG-13', 'Zanzibar', Decimal('21.97')),
 (22, 'R', 'Yangor', Decimal('19.97')),
 (28, 'PG-13', 'Dhaka', Decimal('19.97')),
 (17, 'G', 'Rajkot', Decimal('19.97')),
 (1, 'R', 'Qomsheh', Decimal('19.97')),
 (19, 'PG', 'Najafabad', Decimal('19.96')),
 (30, 'R', 'Fengshan', Decimal('19.95')),
 (21, 'G', 'Wroclaw', Decimal('18.98')),
 (28, 'PG', 'So Leopoldo', Decimal('18.98')),
 (29, 'PG-13', 'Shimoga', Decimal('18.97')),
 (30, 'PG-13', 'Osmaniye', Decimal('18.97')),
 (1, 'NC-17', 'Memphis', Decimal('18.97')),
 (30, 'G', 'Omdurman', Decimal('18.97')),
 (19, 'PG', 'Sokoto', Decimal('18.97')),
 (21, 'PG-13', 'Asuncin', Decimal('18.95')),
 (30, 'PG-13', 'Nagareyama', Decimal('17.98')),
 (21, 'PG-13', 'Parbhani', Decimal('17.98')),
 (19, 'PG', 'Loja', Decimal('17.98')),
 (20, 'PG-13', 'Baha Blanca', Decimal('17.98')),
 (2, 'NC-17', 'Valparai', Decimal('17.98')),
 (18, 'R', 'Sumy', Decimal('17.98')),
 (19, 'G', 'Ozamis', Decimal('17.98')),
 (12, 'NC-17', 'Plock', Decimal('17.98')),
 (23, 'NC-17', 'Adana', Decimal('17.98')),
 (30, 'G', 'Tychy', Decimal('17.98')),
 (17, 'R', 'Birgunj', Decimal('17.97')),
 (30, 'R', 'Robamba', Decimal('17.97')),
 (19, 'R', 'Kowloon and New Kowloon', Decimal('17.97')),
 (19, 'NC-17', 'Lengshuijiang', Decimal('17.97')),
 (30, 'PG-13', 'Tanauan', Decimal('17.96')),
 (17, 'PG-13', 'Ikerre', Decimal('17.95')),
 (10, 'NC-17', 'Vijayawada', Decimal('16.98')),
 (2, 'NC-17', 'Kuwana', Decimal('16.98')),
 (1, 'NC-17', 'Gatineau', Decimal('16.98')),
 (21, 'PG', 'Zhoushan', Decimal('16.98')),
 (11, 'R', 'Gandhinagar', Decimal('16.98')),
 (10, 'G', 'Omdurman', Decimal('16.98')),
 (30, 'PG-13', 'Zhoushan', Decimal('16.98')),
 (17, 'R', 'NDjamna', Decimal('16.98')),
 (9, 'G', 'Fontana', Decimal('16.97')),
 (28, 'R', 'Usolje-Sibirskoje', Decimal('16.97')),
 (30, 'R', 'San Lorenzo', Decimal('16.97')),
 (18, 'NC-17', 'Salzburg', Decimal('16.97')),
 (19, 'NC-17', 'Tangail', Decimal('16.97')),
 (30, 'PG-13', 'Newcastle', Decimal('16.97')),
 (29, 'R', 'Baiyin', Decimal('16.97')),
 (30, 'PG-13', 'Yerevan', Decimal('16.97')),
 (29, 'R', 'Atlixco', Decimal('15.98')),
 (20, 'G', 'Liaocheng', Decimal('15.98')),
 (30, 'G', 'Yinchuan', Decimal('15.98')),
 (17, 'PG-13', 'Yantai', Decimal('15.98')),
 (1, 'PG', 'El Monte', Decimal('15.98')),
 (18, 'R', 'Juazeiro do Norte', Decimal('15.98')),
 (19, 'PG', 'Akron', Decimal('15.98')),
 (18, 'NC-17', 'Ahmadnagar', Decimal('15.98')),
 (28, 'PG', 'Naala-Porto', Decimal('15.98')),
 (30, 'NC-17', 'Vancouver', Decimal('15.98')),
 (30, 'PG-13', 'Santa Rosa', Decimal('15.98')),
 (20, 'PG-13', 'Santa Rosa', Decimal('15.98')),
 (28, 'NC-17', 'Mexicali', Decimal('15.98')),
 (30, 'PG', 'Purnea (Purnia)', Decimal('15.98')),
 (16, 'R', 'Athenai', Decimal('15.98')),
 (30, 'NC-17', 'Quilmes', Decimal('15.98')),
 (1, 'G', 'Songkhla', Decimal('15.98')),
 (6, 'NC-17', 'Phnom Penh', Decimal('15.98')),
 (8, 'PG-13', 'Pontianak', Decimal('15.98')),
 (30, 'R', 'So Bernardo do Campo', Decimal('15.98')),
 (8, 'PG', 'Benin City', Decimal('15.98')),
 (20, 'R', 'Cuman', Decimal('15.98')),
 (27, 'G', 'Salala', Decimal('15.98')),
 (17, 'G', 'Ondo', Decimal('15.98')),
 (29, 'PG', 'Huejutla de Reyes', Decimal('15.97')),
 (1, 'PG-13', 'Moscow', Decimal('15.97')),
 (10, 'PG-13', 'Udaipur', Decimal('15.97')),
 (30, 'PG-13', 'Toulouse', Decimal('15.97')),
 (30, 'NC-17', 'Nakhon Sawan', Decimal('15.97')),
 (30, 'G', 'Hohhot', Decimal('15.97')),
 (17, 'PG', 'Kakamigahara', Decimal('15.97')),
 (18, 'PG', 'Kamarhati', Decimal('15.97')),
 (23, 'PG-13', 'Tiefa', Decimal('15.97')),
 (17, 'NC-17', 'Lancaster', Decimal('15.97')),
 (17, 'G', 'Usak', Decimal('15.97')),
 (20, 'PG-13', 'Johannesburg', Decimal('15.97')),
 (21, 'R', 'Nam Dinh', Decimal('15.96')),
 (19, 'PG-13', 'Touliu', Decimal('14.98')),
 (18

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>day</th>
        <th>rating</th>
        <th>city</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>30</td>
        <td>G</td>
        <td>San Bernardino</td>
        <td>24.97</td>
    </tr>
    <tr>
        <td>30</td>
        <td>NC-17</td>
        <td>Apeldoorn</td>
        <td>23.95</td>
    </tr>
    <tr>
        <td>21</td>
        <td>NC-17</td>
        <td>Belm</td>
        <td>22.97</td>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>Zanzibar</td>
        <td>21.97</td>
    </tr>
    <tr>
        <td>28</td>
        <td>R</td>
        <td>Mwanza</td>
        <td>21.97</td>
    </tr>
</tbody></table></div>

## Slicing

Slicing is the reduction of the dimensionality of a cube by 1 e.g. 3 dimensions to 2, fixing one of the dimensions to a single value. In the example above, we have a 3-dimensional cube on day, rating, and country.

TODO: Write a query that reduces the dimensionality of the above example by limiting the results to only include movies with a `rating` of "PG-13". Again, sort by revenue in descending order and limit to the first 20 rows. The first few rows of your output should match the table below. 

In [10]:
%%time
%%sql

SELECT
    dt.day,
    dm.rating,
    dc.city,
    sum(sales_amount) as revenue
FROM factSales fs 
JOIN dimDate dt ON (dt.date_key = fs.date_key)
JOIN dimCustomer dc ON (dc.customer_key = fs.customer_key)
JOIN dimMovie dm ON (dm.movie_key = fs.movie_key)
WHERE dm.rating = 'PG-13'
GROUP BY dt.day, dm.rating, dc.city
ORDER BY 4 DESC
LIMIT 20

 * postgresql://student:***@127.0.0.1:5432/pagila_star
20 rows affected.
CPU times: user 5.47 ms, sys: 133 µs, total: 5.61 ms
Wall time: 12 ms


[(30, 'PG-13', 'Zanzibar', Decimal('21.97')),
 (28, 'PG-13', 'Dhaka', Decimal('19.97')),
 (30, 'PG-13', 'Osmaniye', Decimal('18.97')),
 (29, 'PG-13', 'Shimoga', Decimal('18.97')),
 (21, 'PG-13', 'Asuncin', Decimal('18.95')),
 (21, 'PG-13', 'Parbhani', Decimal('17.98')),
 (20, 'PG-13', 'Baha Blanca', Decimal('17.98')),
 (30, 'PG-13', 'Nagareyama', Decimal('17.98')),
 (30, 'PG-13', 'Tanauan', Decimal('17.96')),
 (17, 'PG-13', 'Ikerre', Decimal('17.95')),
 (30, 'PG-13', 'Zhoushan', Decimal('16.98')),
 (30, 'PG-13', 'Yerevan', Decimal('16.97')),
 (30, 'PG-13', 'Newcastle', Decimal('16.97')),
 (20, 'PG-13', 'Santa Rosa', Decimal('15.98')),
 (17, 'PG-13', 'Yantai', Decimal('15.98')),
 (30, 'PG-13', 'Santa Rosa', Decimal('15.98')),
 (8, 'PG-13', 'Pontianak', Decimal('15.98')),
 (30, 'PG-13', 'Toulouse', Decimal('15.97')),
 (23, 'PG-13', 'Tiefa', Decimal('15.97')),
 (1, 'PG-13', 'Moscow', Decimal('15.97'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>day</th>
        <th>rating</th>
        <th>city</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>Zanzibar</td>
        <td>21.97</td>
    </tr>
    <tr>
        <td>28</td>
        <td>PG-13</td>
        <td>Dhaka</td>
        <td>19.97</td>
    </tr>
    <tr>
        <td>29</td>
        <td>PG-13</td>
        <td>Shimoga</td>
        <td>18.97</td>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>Osmaniye</td>
        <td>18.97</td>
    </tr>
    <tr>
        <td>21</td>
        <td>PG-13</td>
        <td>Asuncin</td>
        <td>18.95</td>
    </tr>
</tbody></table></div>

## Dicing
Dicing is creating a subcube with the same dimensionality but fewer values for  two or more dimensions. 

TODO: Write a query to create a subcube of the initial cube that includes moves with:
* ratings of PG or PG-13
* in the city of Bellevue or Lancaster
* day equal to 1, 15, or 30

The first few rows of your output should match the table below. 

In [14]:
%%time
%%sql

SELECT
    dt.day,
    dm.rating,
    dc.city,
    sum(fs.sales_amount) as revenue 
FROM factSales fs
JOIN dimDate dt ON (dt.date_key = fs.date_key)
JOIN dimCustomer dc ON (dc.customer_key = fs.customer_key)
JOIN dimMovie dm ON (dm.movie_key = fs.movie_key)
WHERE 
    dm.rating IN ('PG-13', 'PG')
    AND dc.city IN ('Bellevue', 'Lancaster')
    AND dt.day IN ('1','15','30')
GROUP BY dt.day, dm.rating, dc.city
ORDER BY 4 DESC
LIMIT 20

 * postgresql://student:***@127.0.0.1:5432/pagila_star
6 rows affected.
CPU times: user 4.14 ms, sys: 794 µs, total: 4.94 ms
Wall time: 6.49 ms


[(30, 'PG', 'Lancaster', Decimal('12.98')),
 (1, 'PG-13', 'Lancaster', Decimal('5.99')),
 (30, 'PG-13', 'Bellevue', Decimal('3.99')),
 (30, 'PG-13', 'Lancaster', Decimal('2.99')),
 (15, 'PG-13', 'Bellevue', Decimal('1.98')),
 (1, 'PG', 'Bellevue', Decimal('0.99'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>day</th>
        <th>rating</th>
        <th>city</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>30</td>
        <td>PG</td>
        <td>Lancaster</td>
        <td>12.98</td>
    </tr>
    <tr>
        <td>1</td>
        <td>PG-13</td>
        <td>Lancaster</td>
        <td>5.99</td>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>Bellevue</td>
        <td>3.99</td>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>Lancaster</td>
        <td>2.99</td>
    </tr>
    <tr>
        <td>15</td>
        <td>PG-13</td>
        <td>Bellevue</td>
        <td>1.98</td>
    </tr>
</tbody></table></div>